### PR TITLE
DR-593 Provide custodians ability to create jobs in a data project

### PR DESCRIPTION
**Pull request by myessail**

### CODE REVIEW

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java [hunk 1]
```diff
@@ -44,6 +50,9 @@ public StepResult doStep(FlightContext context) {
         Dataset dataset = datasetService.retrieve(datasetId);
         List<String> policyEmails = sam.createDatasetResource(userReq, datasetId);
         bigQueryPdao.grantReadAccessToDataset(dataset, policyEmails);
+        DatasetModel datasetModel = datasetService.retrieveModel(datasetId);
+        System.out.println(datasetModel);
```
Comment (RoriCremer) [inline src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java]: might be worth swapping this to logger if you want to keep this

File: src/test/java/bio/terra/service/iam/AccessTest.java [hunk 2]
```diff
@@ -253,6 +264,56 @@ public void fileAclTest() throws Exception {
...
+
+        // gets the "sample" table and makes a table ref to use in the query
+        String tableRef = BigQueryFixtures.makeTableRef(dataset, dataset.getSchema().getTables().get(1).getName());
+        String sql = String.format("SELECT * FROM %s LIMIT %s", tableRef, 1000);
+        TableResult results = BigQueryFixtures.query(sql, custodianBigQuery);
+        Assert.assertEquals(7, results.getTotalRows());
+
+        assertThat("custodian can access the bq snapshot after it has been shared",
+            custodianHasAccess,
+            equalTo(true));
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/iam/AccessTest.java]: Can you move this check up to just after you define custodianHasAccess?

File: src/test/java/bio/terra/service/iam/AccessTest.java [hunk 3]
```diff
@@ -253,6 +264,56 @@ public void fileAclTest() throws Exception {
...
+
+        SnapshotSummaryModel snapshotSummaryModel =
+            dataRepoFixtures.createSnapshot(custodian(), datasetSummaryModel, "ingest-test-snapshot.json");
```
Comment (marikomedlock) [inline src/test/java/bio/terra/service/iam/AccessTest.java]: Are you testing anything specifically with this snapshot creation? Fine to leave it in, just curious whether there was something in particular you were testing.

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java [hunk 4]
```diff
@@ -44,6 +50,9 @@ public StepResult doStep(FlightContext context) {
         Dataset dataset = datasetService.retrieve(datasetId);
         List<String> policyEmails = sam.createDatasetResource(userReq, datasetId);
         bigQueryPdao.grantReadAccessToDataset(dataset, policyEmails);
+        DatasetModel datasetModel = datasetService.retrieveModel(datasetId);
```
Comment (marikomedlock) [inline src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java]: I think this retrieveModel call causes a duplicate call to retrieve above. If you just need the data project, perhaps you could just call datalocationservice instead?

**Review by marikomedlock — COMMENTED** [review]
Looks good. One comment about possibly eliminating a duplicate call to datasetservice.retrieve.

File: src/test/java/bio/terra/service/iam/AccessTest.java [hunk 5]
```diff
@@ -253,6 +264,56 @@ public void fileAclTest() throws Exception {
...
+        boolean custodianHasAccess =
+            BigQueryFixtures.hasAccess(custodianBigQuery, dataset.getDataProject(), datasetBqSnapshotName);
+
+        // gets the "sample" table and makes a table ref to use in the query
+        String tableRef = BigQueryFixtures.makeTableRef(dataset, dataset.getSchema().getTables().get(1).getName());
+        String sql = String.format("SELECT * FROM %s LIMIT %s", tableRef, 1000);
+        TableResult results = BigQueryFixtures.query(sql, custodianBigQuery);
+        Assert.assertEquals(7, results.getTotalRows());
+
+        assertThat("custodian can access the bq snapshot after it has been shared",
```
Comment (ddietterich) [inline src/test/java/bio/terra/service/iam/AccessTest.java]: The assert can be written more simply as `assertTrue()`

**Review by ddietterich — COMMENTED** [review]
One additional comment beyond the other reviewers.
Looks good to me modulo fixing other review comments.

File: src/test/java/bio/terra/service/iam/AccessTest.java [hunk 6]
```diff
@@ -253,6 +264,56 @@ public void fileAclTest() throws Exception {
...
+        IngestRequestModel request = dataRepoFixtures.buildSimpleIngest(
+            "participant", "ingest-test/ingest-test-participant.json");
+        dataRepoFixtures.ingestJsonData(steward(), datasetId, request);
+
+        request = dataRepoFixtures.buildSimpleIngest(
+            "sample", "ingest-test/ingest-test-sample.json");
+        dataRepoFixtures.ingestJsonData(steward(), datasetId, request);
+
+        DatasetModel dataset = dataRepoFixtures.getDataset(steward(), datasetId);
+        String datasetBqSnapshotName = "datarepo_" + dataset.getName();
```
Comment (RoriCremer) [inline src/test/java/bio/terra/service/iam/AccessTest.java]: Could use PdaoConstant.PDAO_PREFIX for this

**Review by RoriCremer — APPROVED** [review]

File: src/test/java/bio/terra/service/iam/AccessTest.java [hunk 7]
```diff
@@ -91,6 +94,14 @@ private Storage getStorage(String token) {
         return GcsFixtures.getStorage(googleCredentials);
     }
 
+    // create dataset
```
Comment (fboulnois) [inline src/test/java/bio/terra/service/iam/AccessTest.java]: Old comments?

**Review by fboulnois — COMMENTED** [review]
A lot of good suggestions earlier, another short comment:

File: src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java (hunk 4 above)
Comment (myessail) [reply src/main/java/bio/terra/service/dataset/flight/create/CreateDatasetAuthzResource.java]: i looked into it, and the retrieveModel is making a duplicate call -- i just added another retreiveModel that takes in a dataset which skips the duplicate call and just deals with the DataLocationService :)

**Review by fboulnois — APPROVED** [review]
changes lgtm!

**Pull request merged**